File: Python/Histogram Specification.py
import cv2
import numpy as np
import matplotlib.pyplot as plt

# Load input and reference images
input_image = cv2.imread('Enter the path to the image here')
reference_image = cv2.imread('Enter the path to the reference image here')



# Convert to grayscale if needed
if input_image.ndim == 3:
    input_gray = cv2.cvtColor(input_image, cv2.COLOR_BGR2GRAY)
else:
    input_gray = input_image

if reference_image.ndim == 3:
    reference_gray = cv2.cvtColor(reference_image, cv2.COLOR_BGR2GRAY)
else:
    reference_gray = reference_image

# Calculate histograms and CDFs
input_hist, input_bins = np.histogram(input_gray.flatten(), 256, [0, 256])
reference_hist, reference_bins = np.histogram(reference_gray.flatten(), 256, [0, 256])

input_cdf = np.cumsum(input_hist).astype(np.float64)
input_cdf /= input_cdf[-1]  # Normalize to [0,1]

reference_cdf = np.cumsum(reference_hist).astype(np.float64)
reference_cdf /= reference_cdf[-1]

# Create the mapping
mapping = np.zeros(256, dtype=np.uint8)
for i in range(256):
    diff = np.abs(input_cdf[i] - reference_cdf)
    mapping[i] = np.argmin(diff)

# Apply the mapping to get the specified image
specified_image = mapping[input_gray]

# Calculate histogram of specified image
specified_hist, _ = np.histogram(specified_image.flatten(), 256, [0, 256])

# Plotting
plt.figure(figsize=(12, 8))

# Input Image and Histogram
plt.subplot(2, 2, 1)
plt.imshow(input_gray, cmap='gray')
plt.title('Original Grayscale Image')
plt.axis('off')

plt.subplot(2, 2, 2)
plt.bar(np.arange(256), input_hist, width=1)
plt.title('Histogram of Original Image')
plt.xlabel('Gray Level')
plt.ylabel('Frequency')

# Reference Image and Histogram
plt.subplot(2, 2, 3)
plt.imshow(reference_gray, cmap='gray')
plt.title('Reference Grayscale Image')
plt.axis('off')

plt.subplot(2, 2, 4)
plt.bar(np.arange(256), reference_hist, width=1)
plt.title('Histogram of Reference Image')
plt.xlabel('Gray Level')
plt.ylabel('Frequency')

plt.tight_layout()
plt.show()

# Plot the Histogram Specified Image and its Histogram
plt.figure(figsize=(8, 6))

plt.subplot(2, 1, 1)
plt.imshow(specified_image, cmap='gray')
plt.title('Histogram Specified Image')
plt.axis('off')

plt.subplot(2, 1, 2)
plt.bar(np.arange(256), specified_hist, width=1)
plt.title('Histogram of Specified Image')
plt.xlabel('Gray Level')
plt.ylabel('Frequency')

plt.tight_layout()
plt.show()

File: MATLAB/Histogram Specification.m

inputImage = imread('Enter the path to the image here');
referenceImage = imread('Enter the path to the reference image here'); 
if size(inputImage, 3) == 3
    inputGray = rgb2gray(inputImage);
else
    inputGray = inputImage;
end

if size(referenceImage, 3) == 3
    referenceGray = rgb2gray(referenceImage);
else
    referenceGray = referenceImage;
end
[inputCounts, inputGrayLevels] = imhist(inputGray);
[referenceCounts, referenceGrayLevels] = imhist(referenceGray);

inputCDF = cumsum(inputCounts) / numel(inputGray);
referenceCDF = cumsum(referenceCounts) / numel(referenceGray);
mapping = zeros(256, 1, 'uint8');
cdfMap = zeros(256, 1);

for i = 1:256
    [~, idx] = min(abs(inputCDF(i) - referenceCDF));
    mapping(i) = referenceGrayLevels(idx);
end

specifiedImage = mapping(double(inputGray) + 1);
[specifiedCounts, specifiedGrayLevels] = imhist(specifiedImage);
figure;
subplot(2,2,1);
imshow(inputGray);
title('Original Grayscale Image');

subplot(2,2,2);
bar(inputGrayLevels, inputCounts, 'BarWidth', 1);
title('Histogram of Original Image');
xlabel('Gray Level');
ylabel('Frequency');
subplot(2,2,3);
imshow(referenceGray);
title('Reference Grayscale Image');

subplot(2,2,4);
bar(referenceGrayLevels, referenceCounts, 'BarWidth', 1);
title('Histogram of Reference Image');
xlabel('Gray Level');
ylabel('Frequency');
figure;

subplot(2,2,1);
imshow(specifiedImage, []);
title('Histogram Specified Image');

subplot(2,2,2);
bar(specifiedGrayLevels, specifiedCounts, 'BarWidth', 1);
title('Histogram of Specified Image');
xlabel('Gray Level');
ylabel('Frequency');